File: devops/monitoring.js
/**
 * Monitoring - JavaScript
 * 
 * System and application monitoring:
 * - Metrics collection
 * - Alerting
 * - Logging
 */

// Metrics Collector
class MetricsCollector {
    constructor() {
        this.metrics = new Map();
        this.history = [];
    }
    
    record(metricName, value, tags = {}) {
        const timestamp = Date.now();
        const metric = {
            name: metricName,
            value,
            tags,
            timestamp
        };
        
        if (!this.metrics.has(metricName)) {
            this.metrics.set(metricName, []);
        }
        
        this.metrics.get(metricName).push(metric);
        this.history.push(metric);
        
        return metric;
    }
    
    getMetric(metricName, timeRange = null) {
        const metrics = this.metrics.get(metricName) || [];
        
        if (timeRange) {
            const cutoff = Date.now() - timeRange;
            return metrics.filter(m => m.timestamp >= cutoff);
        }
        
        return metrics;
    }
    
    getAverage(metricName, timeRange = null) {
        const metrics = this.getMetric(metricName, timeRange);
        if (metrics.length === 0) return 0;
        
        const sum = metrics.reduce((acc, m) => acc + m.value, 0);
        return sum / metrics.length;
    }
}

// Alert Manager
class AlertManager {
    constructor() {
        this.rules = [];
        this.triggeredAlerts = [];
    }
    
    addRule(rule) {
        this.rules.push(rule);
    }
    
    check(metricsCollector) {
        for (const rule of this.rules) {
            const value = metricsCollector.getAverage(rule.metric, rule.timeRange);
            
            if (this.evaluateRule(rule, value)) {
                this.triggerAlert(rule, value);
            }
        }
    }
    
    evaluateRule(rule, value) {
        switch (rule.operator) {
            case '>':
                return value > rule.threshold;
            case '<':
                return value < rule.threshold;
            case '>=':
                return value >= rule.threshold;
            case '<=':
                return value <= rule.threshold;
            default:
                return false;
        }
    }
    
    triggerAlert(rule, value) {
        const alert = {
            rule: rule.name,
            metric: rule.metric,
            value,
            threshold: rule.threshold,
            severity: rule.severity,
            timestamp: Date.now()
        };
        
        this.triggeredAlerts.push(alert);
        console.log(`🚨 ALERT: ${rule.name} - ${rule.metric} = ${value.toFixed(2)} (threshold: ${rule.threshold})`);
    }
    
    getAlerts() {
        return this.triggeredAlerts;
    }
}

// Logger
class Logger {
    constructor(level = 'info') {
        this.level = level;
        this.logs = [];
        this.levels = { debug: 0, info: 1, warn: 2, error: 3 };
    }
    
    log(level, message, metadata = {}) {
        if (this.levels[level] >= this.levels[this.level]) {
            const logEntry = {
                level,
                message,
                metadata,
                timestamp: new Date().toISOString()
            };
            
            this.logs.push(logEntry);
            console.log(`[${level.toUpperCase()}] ${message}`, metadata);
        }
    }
    
    debug(message, metadata) {
        this.log('debug', message, metadata);
    }
    
    info(message, metadata) {
        this.log('info', message, metadata);
    }
    
    warn(message, metadata) {
        this.log('warn', message, metadata);
    }
    
    error(message, metadata) {
        this.log('error', message, metadata);
    }
    
    getLogs(level = null) {
        if (level) {
            return this.logs.filter(log => log.level === level);
        }
        return this.logs;
    }
}

// Example usage
console.log('=== Monitoring ===\n');

// Metrics
const metrics = new MetricsCollector();
metrics.record('cpu_usage', 45);
metrics.record('cpu_usage', 50);
metrics.record('cpu_usage', 55);
metrics.record('memory_usage', 70);
metrics.record('memory_usage', 75);

console.log('Metrics:');
console.log(`CPU Usage Average: ${metrics.getAverage('cpu_usage').toFixed(2)}%`);
console.log(`Memory Usage Average: ${metrics.getAverage('memory_usage').toFixed(2)}%`);

// Alerts
const alerts = new AlertManager();
alerts.addRule({
    name: 'High CPU',
    metric: 'cpu_usage',
    operator: '>',
    threshold: 50,
    severity: 'warning',
    timeRange: 60000
});

alerts.addRule({
    name: 'High Memory',
    metric: 'memory_usage',
    operator: '>',
    threshold: 70,
    severity: 'critical',
    timeRange: 60000
});

console.log('\nChecking alerts...');
alerts.check(metrics);

// Logging
console.log('\nLogging:');
const logger = new Logger('info');
logger.info('Application started', { port: 3000 });
logger.warn('High memory usage detected', { usage: 75 });
logger.error('Database connection failed', { error: 'Connection timeout' });

module.exports = { MetricsCollector, AlertManager, Logger };

File: devops/monitoring.py
"""
Monitoring - Python

System and application monitoring:
- Metrics collection
- Alerting
- Logging
"""

from typing import Dict, List, Optional
from datetime import datetime
from dataclasses import dataclass
from enum import Enum


class LogLevel(Enum):
    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3


# Metrics Collector
class MetricsCollector:
    def __init__(self):
        self.metrics: Dict[str, List[Dict]] = {}
        self.history: List[Dict] = []
    
    def record(self, metric_name: str, value: float, tags: Dict = None):
        if tags is None:
            tags = {}
        timestamp = datetime.now().timestamp() * 1000
        
        metric = {
            'name': metric_name,
            'value': value,
            'tags': tags,
            'timestamp': timestamp
        }
        
        if metric_name not in self.metrics:
            self.metrics[metric_name] = []
        
        self.metrics[metric_name].append(metric)
        self.history.append(metric)
        
        return metric
    
    def get_metric(self, metric_name: str, time_range: Optional[int] = None) -> List[Dict]:
        metrics = self.metrics.get(metric_name, [])
        
        if time_range:
            cutoff = datetime.now().timestamp() * 1000 - time_range
            return [m for m in metrics if m['timestamp'] >= cutoff]
        
        return metrics
    
    def get_average(self, metric_name: str, time_range: Optional[int] = None) -> float:
        metrics = self.get_metric(metric_name, time_range)
        if not metrics:
            return 0.0
        
        total = sum(m['value'] for m in metrics)
        return total / len(metrics)


# Alert Manager
class AlertManager:
    def __init__(self):
        self.rules: List[Dict] = []
        self.triggered_alerts: List[Dict] = []
    
    def add_rule(self, rule: Dict):
        self.rules.append(rule)
    
    def check(self, metrics_collector: MetricsCollector):
        for rule in self.rules:
            value = metrics_collector.get_average(rule['metric'], rule.get('time_range'))
            
            if self.evaluate_rule(rule, value):
                self.trigger_alert(rule, value)
    
    def evaluate_rule(self, rule: Dict, value: float) -> bool:
        operator = rule['operator']
        threshold = rule['threshold']
        
        if operator == '>':
            return value > threshold
        elif operator == '<':
            return value < threshold
        elif operator == '>=':
            return value >= threshold
        elif operator == '<=':
            return value <= threshold
        return False
    
    def trigger_alert(self, rule: Dict, value: float):
        alert = {
            'rule': rule['name'],
            'metric': rule['metric'],
            'value': value,
            'threshold': rule['threshold'],
            'severity': rule.get('severity', 'warning'),
            'timestamp': datetime.now().timestamp() * 1000
        }
        
        self.triggered_alerts.append(alert)
        print(f"🚨 ALERT: {rule['name']} - {rule['metric']} = {value:.2f} (threshold: {rule['threshold']})")
    
    def get_alerts(self) -> List[Dict]:
        return self.triggered_alerts


# Logger
class Logger:
    def __init__(self, level: str = 'info'):
        self.level = LogLevel[level.upper()]
        self.logs: List[Dict] = []
    
    def log(self, level: str, message: str, metadata: Dict = None):
        if metadata is None:
            metadata = {}
        log_level = LogLevel[level.upper()]
        
        if log_level.value >= self.level.value:
            log_entry = {
                'level': level,
                'message': message,
                'metadata': metadata,
                'timestamp': datetime.now().isoformat()
            }
            
            self.logs.append(log_entry)
            print(f"[{level.upper()}] {message}", metadata)
    
    def debug(self, message: str, metadata: Dict = None):
        self.log('debug', message, metadata)
    
    def info(self, message: str, metadata: Dict = None):
        self.log('info', message, metadata)
    
    def warn(self, message: str, metadata: Dict = None):
        self.log('warn', message, metadata)
    
    def error(self, message: str, metadata: Dict = None):
        self.log('error', message, metadata)
    
    def get_logs(self, level: Optional[str] = None) -> List[Dict]:
        if level:
            return [log for log in self.logs if log['level'] == level]
        return self.logs


# Example usage
if __name__ == "__main__":
    print("=== Monitoring ===\n")
    
    # Metrics
    metrics = MetricsCollector()
    metrics.record('cpu_usage', 45)
    metrics.record('cpu_usage', 50)
    metrics.record('cpu_usage', 55)
    metrics.record('memory_usage', 70)
    metrics.record('memory_usage', 75)
    
    print("Metrics:")
    print(f"CPU Usage Average: {metrics.get_average('cpu_usage'):.2f}%")
    print(f"Memory Usage Average: {metrics.get_average('memory_usage'):.2f}%")
    
    # Alerts
    alerts = AlertManager()
    alerts.add_rule({
        'name': 'High CPU',
        'metric': 'cpu_usage',
        'operator': '>',
        'threshold': 50,
        'severity': 'warning',
        'time_range': 60000
    })
    
    alerts.add_rule({
        'name': 'High Memory',
        'metric': 'memory_usage',
        'operator': '>',
        'threshold': 70,
        'severity': 'critical',
        'time_range': 60000
    })
    
    print("\nChecking alerts...")
    alerts.check(metrics)
    
    # Logging
    print("\nLogging:")
    logger = Logger('info')
    logger.info('Application started', {'port': 3000})
    logger.warn('High memory usage detected', {'usage': 75})
    logger.error('Database connection failed', {'error': 'Connection timeout'})

